### ISSUE TITLE
Debugger attaching too late - Unable to hit breakpoints in Startup.cs

### ISSUE THREAD

**Issue by TroyWitthoeft** [microsoft/vscode-azurefunctions#1971]
In some situations, it looks like the VS Code debugger is unable to attach fast enough to hit specific breakpoints.   For example, if we create a boilerplate C# httptriggered Azure function and add a startup.cs file those "early" breakpoints in the file are often missed.  Example repo - https://github.com/TroyWitthoeft/AzureFunctionsHostStartupBug

Example of missed breakpoint

![76422092-09841600-637b-11ea-863f-e04c2cdf06ef](https://user-images.githubusercontent.com/1102958/76548670-7e337f00-6465-11ea-94ce-37758bbe410d.gif)

Above, why didn't we hit the first breakpoint?
Is there anything we can do to correct this?  I've had quite a few dev's trying to get started with DI in C# using startup.cs and having this issue.   

Note, I am posting here on the suggestion of @WardenGnaw who is assisting me in the csharp extension repo at issue [3644](https://github.com/OmniSharp/omnisharp-vscode/issues/3644).  He [mentioned ]((https://github.com/OmniSharp/omnisharp-vscode/issues/3644#issuecomment-597946902)) that your team might be able to assist.  

Thoughts...  Is it possible to change the `launch.json` file or the `${command:azureFunctions.pickProcess}` command to correct for this?  Is there someway we can signal to the debugger that we are ready for it to attach?

**Comment by EricJizbaMSFT** [comment]
Yeah this is somewhat of a known issue. Similar issues for other languages:
- JavaScript: https://github.com/microsoft/vscode-azurefunctions/issues/1492
- Python: https://github.com/microsoft/vscode-azurefunctions/issues/1841

We're currently blocked on this: https://github.com/Azure/azure-functions-host/issues/4384

**Comment by jsacapdev** [comment]
this happens for my consistently

```
.NET Core SDK (reflecting any global.json):
 Version:   3.1.300
 Commit:    b2475c1295

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  18.04
 OS Platform: Linux
 RID:         ubuntu.18.04-x64
 Base Path:   /usr/share/dotnet/sdk/3.1.300/

Host (useful for support):
  Version: 3.1.4
  Commit:  0090613580

.NET Core SDKs installed:
  3.1.300 [/usr/share/dotnet/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.App 3.1.4 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.4 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
```

vscode
```
Version: 1.45.1
Commit: 5763d909d5f12fe19f215cbfdd29a91c0fa9208a
Date: 2020-05-14T08:27:22.494Z
Electron: 7.2.4
Chrome: 78.0.3904.130
Node.js: 12.8.1
V8: 7.8.279.23-electron.0
OS: Linux x64 5.3.0-1022-azure
```

core tools
```
3.0.2534
```

c#
```
1.22.0
```

azure functions extension
```
0.22.1
```

**Comment by Shilp** [comment]
I am facing the same issue !

**Comment by kevinchatham** [comment]
Same here, has anyone found a fix for this?

**Comment by yutaro113** [comment]
I am also facing same issue.
Would like to know the cause/fix, or at least workaround if any!
Thanks.

**Comment by yutaro113** [comment]
@ejizba Hello, this is just a gentle reminder but is there any plan for fixing this issue? Thank you very much for the confirmation!

**Comment by channeladam** [comment]
Same. Any update?

**Comment by duyn9uyen** [comment]
I've been running in circles trying to figure this out. I'm on a Mac so I don't have an option to use Visual Studio. Anyone found a workaround to hit the breakpoints with VS Code? Adding in Thread.Sleep() didn't seem to help me either.

**Comment by mmulhearn** [comment]
@duyn9uyen [VS for Mac](https://visualstudio.microsoft.com/vs/mac/)

**Comment by brightbeanuser** [comment]
How many years will we need to wait for a fix?

**Comment by nturinski** [comment]
Hi, sorry for the delayed response.  

My understanding is that this PR (https://github.com/microsoft/vscode-azurefunctions/pull/2200) should have fixed this problem, but I can see that it hasn't completely 😞 

What version of the Func Core Tools and Functions Extension are you using?  Does the debugger just miss the breakpoint for the first time and then successfully works after that?  Or does it never even attach?

**Comment by duyn9uyen** [comment]
Func Core Tools: 3.0.3477
Azure Functions (VS Code Extension): v1.6.0

The debugger will not break on startup.cs. It will attach on HTTP requests to the functions themselves. Just not on startup.cs while the app starts.

**Comment by nturinski** [comment]
Thanks for your quick response!

This sounds like it might be a timing issue between when VS Code _actually_ attaches vs. when the process starts. This might be tricky to resolve due to the fact that we are using attach requests for debugging.

> The best way to explain the difference between launch and attach is to think of a launch configuration as a recipe for how to start your app in debug mode before VS Code attaches to it, while an attach configuration is a recipe for how to connect VS Code's debugger to an app or process that's already running. - [Source](https://code.visualstudio.com/docs/editor/debugging#_launch-versus-attach-configurations)

Since the process is required to be running before VS Code can attach, it's most likely missing anything in the `startup.cs` code.  

Perhaps current workarounds are hard-coding a lengthy delay before the breakpoint or configuring the task itself to launch the process.

**Comment by bnchou** [comment]
For info, I can debug Startup.cs with this configuration in launch.json

```
{
  "version": "0.2.0",
  "configurations": [
    {
      "name": "Attach to .NET Functions",
      "type": "coreclr",
      "request": "launch",
      "program": "func",
      "args": ["host", "start"],
      "cwd": "${workspaceFolder}/src/Function/bin/Debug/netcoreapp3.1",
     }
  ]
}
```

**Comment by pkamphuis** [comment]
Neither with no delay, or with long delay. I can't get the breakpoints in the configure code to hit.

**Comment by mvinicius2k** [comment]
Any way to solve this?

Is possible debug manually programming to wait debugger before breakpoints.

```c#
#if DEBUG

while(!Debugger.IsAttached)
    Thread.Sleep(1000);

#endif
``` 

Then with task to start functions project:

```
{
	"label": "func (functions)",
	"type": "shell",
	"command": "func",
	"args": ["host", "start"],
	"group": {
		 "kind": "build",
		 "isDefault": true
	},
	"isBackground": true,
	"options": {
		"cwd": "${workspaceFolder}/<FunctionsProjectFolderHere>",
	}
 }
``` 

And the launch configuration to attach debugger:

```
{
    "name": "Attach Functions",
    "type": "coreclr",
    "request": "attach",
    "preLaunchTask": "func (functions)",          
}
```

The functions project will start and wait the debugger. An error window will apparear because the task is background true and the azure project not started in this moment, you await the worker process initialize, then click in Debug Anyway and finally select the process to attach the debugger.

**Comment by spplante** [comment]
So it's 2026 and this is still a bug?

Created a time trigered azure function with "RunOnStartup=true)" in it's signature so it can launch automatically when doing F5, function executes without hitting a single breakpoint.

If I add a Thread.Sleep(10) at the beginning of my function it does hit the breakpoint, so it confirms it's a timing issue.

Still no fix to debug a function app?
